core/src/triggers.rs

```diff
@@ -8,21 +8,202 @@
 //! search trees (common lisp) or hash tables (racket) to quickly
 //! trigger hooks.
 
-use std::{cmp::min, ops::Deref};
+#![allow(clippy::expect_used, clippy::unwrap_in_result)]
+
+use std::{cmp::min, collections::HashSet, sync::RwLock};
 
 use dashmap::DashMap;
 use iroha_data_model::{
+    domain,
     prelude::*,
     trigger::{self, Action, Repeats, Trigger},
 };
+use tokio::task;
 
 use crate::smartcontracts::{self, FindError, InstructionType, MathError};
 
-/// Specialised structure that maps event filters to Triggers.
-#[derive(Debug, Default, Clone)]
-pub struct TriggerSet(
-    DashMap<trigger::Id, Action>, // TODO: Consider tree structures.
-);
+/// Specialized structure that maps event filters to Triggers.
+#[derive(Debug, Default)]
+pub struct TriggerSet {
+    /// Owns actions
+    all_triggers: DashMap<trigger::Id, Action>,
+    /// Triggers using [`DataEventFilter`]
+    data_triggers: DataTriggerSet,
+    /// Other triggers
+    non_data_triggers: RwLock<HashSet<trigger::Id>>,
+}
+
+/// Set for triggers that uses [`DataEventFilter`]. Allows to quickly find triggers watching for a
+/// specific domain.
+///
+/// Stores only [`trigger::Id`]. All real triggers are stored in [`TriggerSet`].
+/// Then a trigger is removed from `TriggerSet` its id should be removed from this structure too
+#[derive(Debug, Default)]
+struct DataTriggerSet {
+    /// Triggers that are associated with some domain
+    domain_associated: DashMap<domain::Id, HashSet<trigger::Id>>,
+    /// All other data-triggers
+    non_domain_associated: RwLock<HashSet<trigger::Id>>,
+}
+
+impl Clone for DataTriggerSet {
+    fn clone(&self) -> Self {
+        Self {
+            non_domain_associated: RwLock::new(
+                self.non_domain_associated
+                    .read()
+                    .expect("Can't lock actions to clone them")
```

Comment:
Don't panic. Forward the error up the chain.

Reply:
Probably means that you need to replace Clone with something else, .g. `try_clone`.